src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IListConverter.cs

```diff
@@ -23,18 +24,27 @@ protected override void Add(in object? value, ref ReadStack state)
             }
         }
 
-        protected override void CreateCollection(ref Utf8JsonReader reader, scoped ref ReadStack state, JsonSerializerOptions options)
+        private protected override bool TryCreateObject(ref Utf8JsonReader reader, JsonTypeInfo jsonTypeInfo, scoped ref ReadStack state, [NotNullWhen(true)] out object? obj)
         {
-            base.CreateCollection(ref reader, ref state, options);
-            TCollection returnValue = (TCollection)state.Current.ReturnValue!;
-            if (returnValue.IsReadOnly)
+            if (base.TryCreateObject(ref reader, jsonTypeInfo, ref state, out obj))
             {
-                state.Current.ReturnValue = null; // clear out for more accurate JsonPath reporting.
-                ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(TypeToConvert, ref reader, ref state);
+                TCollection returnValue = (TCollection)obj;
+
+                if (returnValue.IsReadOnly)
+                {
+                    ThrowHelper.ThrowNotSupportedException_CannotPopulateCollection(TypeToConvert, ref reader, ref state);
+                }
+
+                return true;
```

Comment:
we could provide generic implementation of this for all collections (i.e. bool IsReadonly(TCollection obj))